proxy/proxyserver/prefetch.go

```diff
@@ -0,0 +1,268 @@
+	metrics       tally.Scope
+}
+
+// Request and response payloads.
+type prefetchBody struct {
+	Tag     string `json:"tag"`
+	TraceId string `json:"trace_id"`
+}
+
+type prefetchResponse struct {
+	Tag        string `json:"tag"`
+	Prefetched bool   `json:"prefetched"`
+	Status     string `json:"status"`
+	Message    string `json:"message"`
+	TraceId    string `json:"trace_id"`
+}
+
+type prefetchError struct {
+	Error      string `json:"error"`
+	Prefetched bool   `json:"prefetched"`
+	Status     string `json:"status"`
+	Message    string `json:"message"`
+	TraceId    string `json:"trace_id,omitempty"`
+}
+
+// NewPrefetchHandler constructs a new PrefetchHandler.
+func NewPrefetchHandler(client blobclient.ClusterClient, tagClient tagclient.Client, metrics tally.Scope) *PrefetchHandler {
+	return &PrefetchHandler{
+		clusterClient: client,
+		tagClient:     tagClient,
+		metrics:       metrics.SubScope("prefetch"),
+	}
+}
+
+// newPrefetchSuccessResponse constructs a successful response.
+func newPrefetchSuccessResponse(tag, msg, traceId string) *prefetchResponse {
+	return &prefetchResponse{
+		Tag:        tag,
+		Prefetched: true,
+		Status:     StatusSuccess,
+		Message:    msg,
+		TraceId:    traceId,
+	}
+}
+
+// newPrefetchError constructs an error response.
+func newPrefetchError(status int, msg, traceId string) *prefetchError {
+	return &prefetchError{
+		Error:      http.StatusText(status),
+		Prefetched: false,
+		Status:     StatusFailure,
+		Message:    msg,
+		TraceId:    traceId,
+	}
+}
+
+// writeJSON writes the JSON payload with the given HTTP status.
+func writeJSON(w http.ResponseWriter, status int, payload interface{}) {
+	w.WriteHeader(status)
+	json.NewEncoder(w).Encode(payload)
+}
+
+func writeBadRequestError(w http.ResponseWriter, msg, traceId string) {
+	writeJSON(w, http.StatusBadRequest, newPrefetchError(http.StatusBadRequest, msg, traceId))
+}
+
+func writeInternalError(w http.ResponseWriter, msg, traceId string) {
+	writeJSON(w, http.StatusInternalServerError, newPrefetchError(http.StatusInternalServerError, msg, traceId))
+}
+
+func writePrefetchResponse(w http.ResponseWriter, tag, msg, traceId string) {
+	writeJSON(w, http.StatusOK, newPrefetchSuccessResponse(tag, msg, traceId))
+}
+
+// Handle processes the prefetch request.
+func (ph *PrefetchHandler) Handle(w http.ResponseWriter, r *http.Request) {
+	ph.metrics.Counter("requests").Inc(1)
+	var reqBody prefetchBody
+	if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
+		writeBadRequestError(w, fmt.Sprintf("failed to decode request body: %s", err), "")
+		log.With("error", err).Error("Failed to decode request body")
+		return
+	}
+	logger := log.With("trace_id", reqBody.TraceId)
+
+	namespace, tag, err := parseTag(reqBody.Tag)
+	if err != nil {
+		writeBadRequestError(w, fmt.Sprintf("tag: %s, invalid tag format: %s", reqBody.Tag, err), reqBody.TraceId)
+		return
+	}
+
+	tagRequest := fmt.Sprintf("%s%%2F%s", namespace, tag)
```

Comment:
Consider using https://pkg.go.dev/net/url#QueryEscape query escape instead of hardcoding